packages/ref-core/src/cmip_ref_core/constraints.py

```diff
@@ -140,6 +140,82 @@ def validate(self, group: pd.DataFrame) -> bool:
         return all(value in group[self.dimension].values for value in self.required_facets)
 
 
+@frozen
+class RequireContiguousTimerange:
+    """
+    A constraint that requires datasets to have a contiguous timerange.
+    """
+
+    group_by: list[str]
+    """
+    The fields to group the datasets by. Each group must be contiguous in time
+    to fulfill the constraint.
+    """
+
+    def validate(self, group: pd.DataFrame) -> bool:
+        """
+        Check that all subgroups of the group have a contiguous timerange.
+        """
+        # Maximum allowed time difference between the end of one file and the
+        # start of the next file.
+        max_timedelta = pd.Timedelta(
+            days=31,  # Maximum number of days in a month.
+            hours=1,  # Allow for potential rounding errors.
+        )
+        group = group.dropna(subset=["start_time", "end_time"])
+        if len(group) < 2:  # noqa: PLR2004
+            return True
+
+        for _, subgroup in group.groupby(self.group_by):
+            if len(subgroup) < 2:  # noqa: PLR2004
+                continue
+            sorted_group = subgroup.sort_values("start_time")
+            start_series = sorted_group["start_time"]
+            end_series = sorted_group["end_time"]
+            # Sometimes the elements of start_series.values are of type datetime64[ns]
+            # and sometimes its elements are of type datetime.datetime.
+            # Convert both arrays to datetime.datetime objects to make sure they
+            # can be subtracted.
+            if hasattr(start_series, "dt"):
+                start_array = start_series.dt.to_pydatetime()
+            else:
+                start_array = start_series.values  # type: ignore[assignment]
```

Comment:
@lewisjared I'm not very familiar with pandas, so I'm not sure how to generate a test case for this. However, this line is triggered by running `pytest packages/ref/tests/unit/test_solver.py::test_solve_metrics_default_solver`.